google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/functions/ResetWindowsPassword.java

```diff
@@ -0,0 +1,225 @@
+ * limitations under the License.
+ */
+package org.jclouds.googlecomputeengine.compute.functions;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.math.BigInteger;
+import java.nio.charset.Charset;
+import java.security.InvalidKeyException;
+import java.security.KeyFactory;
+import java.security.KeyPair;
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.RSAPublicKeySpec;
+import java.text.SimpleDateFormat;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.TimeZone;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import javax.annotation.Resource;
+import javax.crypto.BadPaddingException;
+import javax.crypto.Cipher;
+import javax.crypto.IllegalBlockSizeException;
+import javax.crypto.NoSuchPaddingException;
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import com.google.gson.GsonBuilder;
+import org.jclouds.compute.reference.ComputeServiceConstants;
+import org.jclouds.crypto.Crypto;
+import org.jclouds.googlecomputeengine.GoogleComputeEngineApi;
+import org.jclouds.googlecomputeengine.domain.Instance;
+import org.jclouds.googlecomputeengine.domain.Instance.SerialPortOutput;
+import org.jclouds.googlecomputeengine.domain.Metadata;
+import org.jclouds.googlecomputeengine.domain.Operation;
+import org.jclouds.googlecomputeengine.features.InstanceApi;
+import org.jclouds.logging.Logger;
+
+import com.google.common.base.Function;
+import com.google.common.base.Predicate;
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.io.BaseEncoding;
+import com.google.common.util.concurrent.Atomics;
+import com.google.gson.Gson;
+import org.jclouds.util.Predicates2;
+
+/**
+ * References:
+ * <ul>
+ *   <li><a href="https://cloud.google.com/compute/docs/instances/automate-pw-generation">automate-pw-generation</a>
+ *   <li><a href="https://github.com/GoogleCloudPlatform/compute-image-windows/blob/master/examples/windows_auth_java_sample.java">windows_auth_java_sample.java</a>
+ * </ul>
+ * 
+ * In brief, the sequence is:
+ * <ol>
+ *   <li>Generate a temporary key for encrypting and decrypting the password
+ *   <li>Send the RSA public key to the instance, by settings its metadata
+ *   <li>Retrieve the result from the {@link SerialPortOutput}
+ *   <li>Decode and decrypt the result.
+ * </ol>
+ */
+public class ResetWindowsPassword implements Function<Map<String, ?>, String> {
+
+   /**
+    * Indicates when the key should expire. Keys are one-time use, so the metadata doesn't need to stay around for long.
+    * 10 minutes chosen to allow for differences between time on the client
+    * and time on the server.
+    */
+   private static final long EXPIRE_DURATION = 10 * 60 * 1000;
+   private String userName;
+   private String email;
```

Comment:
I see now. Instead of making it member variables, please, make it local variables in the apply method and pass them to the methods that need them. We should avoid mutable state wherever possible.